File: programmers/LV2/리코쳇로봇.cpp
#include <string>
#include <vector>
#include <queue>
#include <utility>
#include <iostream>

using namespace std;

typedef pair<int, int> Point;

vector<string> boardArr;
int n, m;
int vis[102][102] = {-1, };

Point move(int dir, Point start) {
    Point position = start;
    if(dir == 0) { // UP
        if(position.first == 0) { return position; }
        while(boardArr[position.first][position.second] != 'D') {
            if(position.first <= 0) break;
            position.first--;
        }
        if(position.first == 0 && boardArr[position.first][position.second] == 'D') {
            position.first++;
        } else if(position.first != 0) {
            position.first++;
        }
        return position;        
    } else if(dir == 1) { // DOWN
        if(position.first == n - 1) { return position; }
         while(boardArr[position.first][position.second] != 'D') {
            if(position.first >= n - 1) break;
            position.first++;
        }
        if(position.first == n - 1 && boardArr[position.first][position.second] == 'D') {
            position.first--;
        } else if(position.first != n - 1) {
            position.first--;
        }
        return position;
    } else if(dir == 2) { // LEFT
        if(position.second == 0) { return position; }
        while(boardArr[position.first][position.second] != 'D') {
            if(position.second <= 0) break;
            position.second--;
        }
        if(position.second == 0 && boardArr[position.first][position.second] == 'D') {
            position.second++;
        } else if(position.second != 0) {
            position.second++;
        }
        return position;
    } else { // RIGHT
        if(position.second == m - 1) { return position; }
        while(boardArr[position.first][position.second] != 'D') {
            if(position.second >= m - 1) break;
            position.second++;
        }
        if(position.second == m - 1 && boardArr[position.first][position.second] == 'D') {
            position.second--;
        } else if(position.second != m - 1) {
            position.second--;
        }
        return position;
    }
}

int solution(vector<string> board) {
    boardArr = board;
    n = boardArr.size();
    m = boardArr[0].size();
    for(int y = 0; y < n; y++) {
        for(int x = 0; x < m; x++) {
            vis[y][x] = -1;
        }
    }
    
    Point start;
    Point end;
    for(int y = 0; y < n; y++) {
        for(int x = 0; x < m; x++) {
            if(boardArr[y][x] == 'R') {
                start = {y, x};
            }
            if(boardArr[y][x] == 'G') {
                end = {y, x};
            }
        }
    }
    
    queue<Point> q;
    q.push(start);
    vis[start.first][start.second] = 0;
    
    while(!q.empty()) {
        Point cur = q.front();
        q.pop();
        
        for(int dir = 0; dir < 4; dir++) {
            Point dPosition = move(dir, cur);
            int ny = dPosition.first;
            int nx = dPosition.second;
            
            if(ny < 0 || ny >= n || nx < 0 || nx >= m) continue;
            if(vis[ny][nx] != -1) continue;
            
            q.push(dPosition);
            vis[ny][nx] = vis[cur.first][cur.second] + 1;
        }
    }
    
    return vis[end.first][end.second];
}

File: programmers/LV2/리코쳇로봇.swift
import Foundation

typealias Position = (y: Int, x: Int)

public struct Queue<T> {
    fileprivate var array = [T?]()
    fileprivate var head = 0
    
    public var isEmpty: Bool {
        return count == 0
    }
    
    public var count: Int {
        return array.count - head
    }
    
    public mutating func enqueue(_ element: T) {
        array.append(element)
    }
    
    public mutating func dequeue() -> T? {
        guard head < array.count, let element = array[head] else { return nil }
        
        array[head] = nil
        head += 1
        
        let percentage = Double(head) / Double(array.count)
        if array.count > 50 && percentage > 0.25 {
            array.removeFirst(head)
            head = 0
        }
        
        return element
    }
}

enum Dir: CaseIterable {
    case up
    case down
    case left
    case right
}

var boardArr: [[String]] = []

func move(_ dir: Dir, _ start: Position) -> Position {
    var position = start
    
    switch dir {
    case .up:
        if position.y == 0 { return position }
        while boardArr[position.y][position.x] != "D" {
            if position.y <= 0 { break }
            position.y -= 1
        }
        if boardArr[position.y][position.x] == "D" && position.y == 0 {
            position.y += 1
        } else if position.y != 0  {
            position.y += 1
        }
        return position
    case .down:
        if position.y == boardArr.count - 1 { return position }
        while boardArr[position.y][position.x] != "D" {
            if position.y >= boardArr.count - 1 { break }
            position.y += 1
        }
        if boardArr[position.y][position.x] == "D" && position.y == boardArr.count - 1 {
            position.y -= 1
        } else if position.y != boardArr.count - 1 {
            position.y -= 1
        }
        return position
    case .left:
        if position.x == 0 { return position }
        while boardArr[position.y][position.x] != "D" {
            if position.x <= 0 { break }
            position.x -= 1
        }
        if boardArr[position.y][position.x] == "D" && position.x == 0 {
            position.x += 1
        } else if position.x != 0  {
            position.x += 1
        }
        return position
    case .right:
        if position.x == boardArr[0].count - 1 { return position }
        while boardArr[position.y][position.x] != "D" {
            if position.x >= boardArr[0].count - 1 { break }
            position.x += 1
        }
        if boardArr[position.y][position.x] == "D" && position.x == boardArr[0].count - 1 {
            position.x -= 1
        } else if position.x != boardArr[0].count - 1 {
            position.x -= 1
        }
        return position
    }
}

func solution(_ board: [String]) -> Int {
    boardArr = board.map { $0.map { String($0) } }
    
    var start: Position = (0, 0)
    var end: Position = (0, 0)
    for y in 0..<boardArr.count {
        for x in 0..<boardArr[0].count {
            if boardArr[y][x] == "R" {
                start = (y, x)
            }
            if boardArr[y][x] == "G" {
                end = (y, x) 
            }
        }
    }
    
    var vis: [[Int]] = Array(
        repeating: Array(repeating: -1, count: boardArr[0].count), 
        count: boardArr.count
    ) 
    var queue = Queue<Position>()
    
    queue.enqueue(start)
    vis[start.y][start.x] = 0
    
    while !queue.isEmpty {
        let cur = queue.dequeue()!
        
        for dir in Dir.allCases {
            let dPosition = move(dir, cur)
            let dy = dPosition.y
            let dx = dPosition.x
            
            if dy < 0 || dy >= boardArr.count || dx < 0 || dx >= boardArr[0].count { continue }
            if vis[dy][dx] != -1 { continue }
            
            queue.enqueue(dPosition)
            vis[dy][dx] = vis[cur.y][cur.x] + 1
        }
    }
    
    return vis[end.y][end.x]
}
